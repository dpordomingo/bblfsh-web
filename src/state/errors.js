const initialState = [];

export const ADD = 'bblfsh/errors/ADD';
export const SET = 'bblfsh/errors/SET';
export const REMOVE = 'bblfsh/errors/REMOVE';
export const CLEAR = 'bblfsh/errors/CLEAR';

export const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ADD:
      return state.concat(action.errors || []);
    case SET:
      return action.errors || [];
    case REMOVE:
      return [...state.slice(0, action.idx), ...state.slice(action.idx + 1)];
    case CLEAR:
      return [];
    default:
      return state;
  }
};

export const add = errors => ({
  type: ADD,
  errors,
});

export const set = errors => ({
  type: SET,
  errors,
});

export const remove = idx => ({
  type: REMOVE,
  idx,
});

export const clear = () => ({
  type: CLEAR,
});
