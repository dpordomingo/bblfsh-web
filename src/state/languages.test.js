import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';

const mockStore = configureMockStore([thunk]);

import {
  reducer,
  initialState,
  LOADING,
  LOADED,
  LOAD_FAILED,
  SET,
  SELECT,
  load,
  select,
  getLanguageMode,
} from './languages';
import { add as errorsAdd } from './errors';

describe('languages/reducer', () => {
  it('LOADING', () => {
    expect(reducer(initialState, { type: LOADING })).toMatchSnapshot();
  });

  it('LOADED', () => {
    expect(
      reducer(
        { ...initialState, loading: true },
        {
          type: LOADED,
          languages: { python: { name: 'python' } },
        }
      )
    ).toMatchSnapshot();
  });

  it('LOAD_FAILED', () => {
    expect(
      reducer({ ...initialState, loading: true }, { type: LOAD_FAILED })
    ).toMatchSnapshot();
  });

  it('SET', () => {
    expect(
      reducer(initialState, { type: SET, actual: 'python' })
    ).toMatchSnapshot();
  });

  it('SELECT', () => {
    expect(
      reducer(initialState, { type: SELECT, selected: 'python' })
    ).toMatchSnapshot();
  });
});

describe('languages/actions', () => {
  it('load', () => {
    fetch.mockResponse(
      JSON.stringify([
        { id: 'python', name: 'Python', url: 'python-driver' },
        { id: 'java', name: 'Java', url: 'java-driver' },
      ])
    );

    const store = mockStore({ languages: initialState });
    const expectedActions = [
      { type: LOADING },
      {
        type: LOADED,
        languages: {
          java: { mode: 'text/x-java', name: 'Java', url: 'java-driver' },
          python: { mode: 'python', name: 'Python', url: 'python-driver' },
        },
      },
    ];

    return store.dispatch(load()).then(() => {
      expect(store.getActions()).toEqual(expectedActions);
    });
  });

  it('load failed', () => {
    fetch.mockReject();

    const store = mockStore({ languages: initialState });
    const expectedActions = [
      { type: LOADING },
      { type: LOAD_FAILED },
      errorsAdd(['Unable to load the list of available drivers.']),
    ];

    return store.dispatch(load()).then(() => {
      expect(store.getActions()).toEqual(expectedActions);
    });
  });

  it('select', () => {
    const store = mockStore({
      languages: {
        ...initialState,
        languages: {
          ...initialState.languages,
          python: { name: 'python' },
        },
      },
    });

    store.dispatch(select('python'));
    expect(store.getActions()).toEqual([
      {
        type: SELECT,
        selected: 'python',
      },
    ]);

    store.clearActions();

    store.dispatch(select('unknown'));
    expect(store.getActions()).toEqual([
      {
        type: SELECT,
        selected: '',
      },
    ]);
  });
});

describe('getLanguageMode', () => {
  it('has selected language', () => {
    const mode = getLanguageMode({
      languages: {
        languages: {
          java: { mode: 'x-java' },
        },
        selected: 'java',
        actual: 'java',
      },
    });
    expect(mode).toEqual('x-java');
  });

  it('has actual language', () => {
    const mode = getLanguageMode({
      languages: {
        languages: {
          java: { mode: 'x-java' },
        },
        selected: '',
        actual: 'java',
      },
    });
    expect(mode).toEqual('x-java');
  });

  it('unknown language', () => {
    const mode = getLanguageMode({
      languages: {
        languages: {
          java: { mode: 'x-java' },
        },
        selected: '',
        actual: '',
      },
    });
    expect(mode).toEqual('');
  });
});
