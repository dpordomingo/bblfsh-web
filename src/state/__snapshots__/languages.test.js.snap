// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`languages/reducer LOAD_FAILED 1`] = `
Object {
  "actual": "",
  "languages": Object {
    "": Object {
      "name": "(auto)",
    },
  },
  "loading": false,
  "selected": "",
}
`;

exports[`languages/reducer LOADED 1`] = `
Object {
  "actual": "",
  "languages": Object {
    "": Object {
      "name": "(auto)",
    },
    "python": Object {
      "name": "python",
    },
  },
  "loading": false,
  "selected": "",
}
`;

exports[`languages/reducer LOADING 1`] = `
Object {
  "actual": "",
  "languages": Object {
    "": Object {
      "name": "(auto)",
    },
  },
  "loading": true,
  "selected": "",
}
`;

exports[`languages/reducer SELECT 1`] = `
Object {
  "actual": "python",
  "languages": Object {
    "": Object {
      "name": "(auto)",
    },
  },
  "loading": true,
  "selected": "python",
}
`;

exports[`languages/reducer SET 1`] = `
Object {
  "actual": "python",
  "languages": Object {
    "": Object {
      "name": "(auto)",
    },
  },
  "loading": true,
  "selected": "",
}
`;
